### PR TITLE
cli: add option to export connection envvars

### LINKED ISSUE #15: cli: simplify configuration
```console
$ docker-services-cli --version
docker-services-cli, version 0.2.1
```

## Describe the bug

If one misses setting for example the `SQLALCHEMY_DATABASE_URI`, the tests will just run using the default DB (which in the case of Invenio is SQLite [see docs](https://invenio-db.readthedocs.io/en/latest/configuration.html#SQLALCHEMY_DATABASE_URI)), e.g.:

```console
$ docker-services-cli up postgresql
$ pytest -m tests
$ # pytest/application will connect to the default DB
$ # test can pass but they are not run using postgresql
```

## Expected behavior

```console
$ export DB=MYSQL
$ docker-services-cli up
$ # injects the needed env variables, e.g. sqlalchemy connection string
$ env | grep SQLALCHEMY_DATABASE_URI
SQLALCHEMY_DATABASE_URI=<mysql-container-connection-string->
$ # run python tests, tests/app code read connection strings from environment 
$ docker-services-cli down
$ # cleans up environment
$ env | grep SQLALCHEMY_DATABASE_URI | wc -l
0
```

## Additional context

>Yep. Iit simplifies both GH Actions and running tests locally, so I would go ahead.

>It could also be:

>```
>export DB=MYSQL_5_LATEST
>docker-services-cli up
>```

>If you want the latest release:

>```
>export DB=MYSQL
>docker-services-cli up
>```

_Originally posted by @lnielsen in https://github.com/inveniosoftware/invenio-records/pull/239#discussion_r534083525_

### PR DESCRIPTION
**Pull request by diegodelemos**
* Changes config splitting default minor versions and exportable
  environment variables per service.

* Adds new flag `--env` so developers can eval the output and
  configure their shell (closes #15).

#### Local test scenarios:

The solution doesn't implement exactly the description in #15, because it is not possible to change the parent shell environment from Python. Instead, the solution which most of the CLI programs implement, having a command/option that is directly exportable, has been implemented.

Scenario 1, short way:
```console
$ env | grep SQL
$ eval "$(docker-services-cli up postgresql12 elasticsearch6 --env)"
$ env | grep SQL
SQLALCHEMY_DATABASE_URI=postgresql+psycopg2://johndoe:password@localhost:5432/database
$ docker ps | grep postgres:12
7ffdd557a950   postgres:12.4                                              "docker-entrypoint.s…"   About a minute ago   Up About a minute   0.0.0.0:5432->5432/tcp                           docker_services_cli_postgresql_1
$ docker ps | grep elasticsearch-oss:6.8.12 
346c7f1adca0   docker.elastic.co/elasticsearch/elasticsearch-oss:6.8.12   "/usr/local/bin/dock…"   55 seconds ago   Up 54 seconds   0.0.0.0:9200->9200/tcp, 0.0.0.0:9300->9300/tcp   docker_services_cli_elasticsearch_1
$ eval "$(docker-services-cli down --env)"
$ env | grep SQL
```

Scenario 2, the first time developer sees the feature:

```console
$ # Clean env
$ docker-services-cli down
$ unset SQLALCHEMY_DATABASE_URI
$ docker-services-cli up postgresql12 elasticsearch6 --env
: '
Environment setup
Creating network "docker_services_cli_default" with the default driver
Creating docker_services_cli_postgresql_1    ... done
Creating docker_services_cli_elasticsearch_1 ... done
postgresql not ready at 1 retries, waiting 2s
postgresql up and running!
elasticsearch not ready at 1 retries, waiting 2s
elasticsearch not ready at 2 retries, waiting 4s
elasticsearch up and running!
Services up!
'
export SQLALCHEMY_DATABASE_URI=postgresql+psycopg2://johndoe:password@localhost:5432/database
# Configure your environment running:
# eval "$(docker-services-cli up postgresql elasticsearch --env)"
$ env | grep SQL
$ eval "$(docker-services-cli setup-env postgresql elasticsearch)"
$ env | grep SQL
SQLALCHEMY_DATABASE_URI=postgresql+psycopg2://johndoe:password@localhost:5432/database
$ docker-services-cli down --env
: '
Stopping docker_services_cli_postgresql_1    ... done
Stopping docker_services_cli_elasticsearch_1 ... done
Removing docker_services_cli_postgresql_1    ... done
Removing docker_services_cli_elasticsearch_1 ... done
Removing network docker_services_cli_default
Services down!
'
unset SQLALCHEMY_DATABASE_URI=postgresql+psycopg2://johndoe:password@localhost:5432/database
unset SQLALCHEMY_DATABASE_URI=mysql+pymysql://johndoe:password@localhost:3306/database
unset CACHE_TYPE=redis
unset CACHE_TYPE=memcached
unset BROKER_URL=redis://localhost:6379/0
unset BROKER_URL=amqp://localhost:5672//
# Configure your environment running:
# eval "$(docker-services-cli down elasticsearch postgresql mysql redis memcached rabbitmq --env)"
$ eval "$(docker-services-cli down elasticsearch postgresql mysql redis memcached rabbitmq --env)"
$ env | grep SQL
```

**Note**: `docker-services-cli` doesn't take services as a parameter, so the command will output `unset` instructions for all possible environment variables, even the ones that were not set. 

Scenario 3, backwards compatible way:

```console
$ # Clean env
$ docker-services-cli down
$ docker-services-cli up postgresql
$ docker ps | grep -i postgres:9
7524edf202a5   postgres:9.6.19   "docker-entrypoint.s…"   23 seconds ago   Up 22 seconds   0.0.0.0:5432->5432/tcp   docker_services_cli_postgresql_1
```

### Questions
- Should `--env` be the default output of `up/down` commands? 

⚠️  Documentation not updated until we agree on a final solution.

### CODE REVIEW

File: docker_services_cli/env.py
Comment (diegodelemos) [inline docker_services_cli/env.py:98]: 👨‍💻 Changing this to fail instead of just showing a warning. Doing things like `docker-services-cli up postgres111` with the current code would result in running tests with version 9 just because of a typo...

**Comment by diegodelemos** [comment]
Feedback welcome. Usage in [PR description](https://github.com/inveniosoftware/docker-services-cli/pull/16#issue-531688126) core change in https://github.com/inveniosoftware/docker-services-cli/pull/16/commits/32f7b0c800cc283b6380f4dcd9de38e30dde8afa.

cc @lnielsen @ppanero @mvidalgarcia

File: docker_services_cli/config.py
Comment (ppanero) [inline docker_services_cli/config.py:52]: minor: you are probably aware but there are many repos that are already configured with `invenio:invenio` on the GA CI. This will affect those.

File: docker_services_cli/docker-services.yml
Comment (ppanero) [inline docker_services_cli/docker-services.yml:35]: comment: we were using `es` to be consistent with invenio (e.g. with cookiecutter-invenio-instance and co). That said, I also prefer `elasticsearch`.

File: docker_services_cli/config.py
Comment (ppanero) [inline docker_services_cli/config.py:85]: moderate: here it is named `...MINOR...`. However, I see no `MAJOR` definition anywhere and even [below](https://github.com/inveniosoftware/docker-services-cli/pull/16/files#diff-1fc38c9800c4622657c57dc615e70a5f0ff866314547364b8338fdef5a116d86R100) major versions are taken from the `minor` lists. Should it be simple "versions"

File: setup.py
Comment (ppanero) [inline setup.py:23]: minor: do we need this upper bound?

**Review by ppanero — COMMENTED** [review]
Overall it LGTM. It is a really nice job and makes some things more consistent (thanks for all the ' vs ", and line adjustments too!)

I am not a fan of having to use `eval` and add complexity to the way the command is used. It makes is highly error prone since not all terminals behave the same way, etc. So in my opinion it has been:

- Simplified at configuration level
- Made harder at "running it" level

I was away and unsure of the decisions 😅 so I limit myself to check the code functionality.

One last *minor* thing: shuoldn't the `unset` printed in console be simply the variable name? e.g. `unset SQLALCHEMY_DATABASE_URI`, instead of `unset unset SQLALCHEMY_DATABASE_URI=...`?  Otherwise, at least in zsh, it fails (an e.g. my comment of making it harder to run).

```
~ export TEST=thisisatest
~ unset TEST=thisisatest
unset: TEST=thisisatest: invalid parameter name
~ unset TEST
```

File: setup.py
Comment (diegodelemos) [reply setup.py:23]: Since `pytest-invenio` is the central repo to define test dependencies, I just mirrored [what it currently has in its `setup.py`](https://github.com/inveniosoftware/pytest-invenio/blob/798eaa4140dcdfe6e2cb51338915701a993545ee/setup.py#L58) (Note that `pytest-invenio` depends on `docker-services-cli` that's why the duplication).

File: docker_services_cli/config.py
Comment (diegodelemos) [reply docker_services_cli/config.py:85]: I am fine with changing it back, it just helped me to understand/read things better while implementing this change.

**Comment by mvidalgarcia** [comment]
I confirm the `unset` problem mentioned by @ppanero:

```console
$ eval "$(docker-services-cli down --env)"             
Stopping docker_services_cli_postgresql_1 ... done
Removing docker_services_cli_postgresql_1 ... done
Removing network docker_services_cli_default
unset: SQLALCHEMY_DATABASE_URI=postgresql+psycopg2://johndoe:password@localhost:5432/database: invalid parameter name
```

File: docker_services_cli/docker-services.yml
Comment (diegodelemos) [reply docker_services_cli/docker-services.yml:35]: To put it simply:
- `es` is part of `postgresql`: having it named `es` is a major blocker when implementing `override_default_env`. 
- `es` doesn't follow the naming scheme of the rest of services in the docker compose file  or anywhere in the repo. All services are called full name, `es` is short.

File: docker_services_cli/config.py
Comment (diegodelemos) [reply docker_services_cli/config.py:52]: This will definitely break the current GA CI builds because they have `SQLALCHEMY_DATABASE_URI` hardcoded. However, that is actually something we want to fix with this PR because now `SQLALCHEMY_DATABASE_URI` will be controlled by `docker-services-cli` instead of being written in every Invenio repo's GA CI `tests.yml`.

My guiding principle was that I heard we wanted to keep `docker-services-cli` generic, without Invenio dependencies. Setting name, password and db name to `invenio` seems to go against that principle.

Said that, we can leave it like it is, and once we merge and release this tool, we can progressively remove `SQLALCHEMY_DATABASE_URI` from everywhere, and only then come and change this, what do you think?

**Comment by diegodelemos** [comment]
> (thanks for all the ' vs ", and line adjustments too!)

' vs " -> I have `black` on save
line adjustments too -> most of them have come from pycodestyle checks.

> One last minor thing: shuoldn't the unset printed in console be simply the variable name? e.g. unset SQLALCHEMY_DATABASE_URI, instead of unset unset SQLALCHEMY_DATABASE_URI=...? Otherwise, at least in zsh, it fails (an e.g. my comment of making it harder to run).

Definitely 🤦‍♀️ , fixing it.

**Comment by diegodelemos** [comment]
> I am not a fan of having to use eval and add complexity to the way the command is used. It makes is highly error prone since not all terminals behave the same way, etc. So in my opinion it has been:

>- Simplified at configuration level
>- Made harder at "running it" level

1. We have a proven usability problem, everyone in GA actions sprint got confused and got it wrong, so it needs to be fixed 100%.
2. First proposed solution is not technically possible in a proper way (injecting envvars to parent processes).
3. This is presented as an opt-in feature to solve the problem, if one leaves out `--env` then there is no complication, other than you have to remember yourself to set the envvars (original problem).
  i. To be 100% clear, I am not attached in any way to this, I just followed what is a non written standard in CLI tools when they need to export configuration to be used by other apps. If you have a better idea on how to solve this so it is easier for everyone I am all ears 😊 .

File: docker_services_cli/docker-services.yml
Comment (ppanero) [reply docker_services_cli/docker-services.yml:35]: ohh damn! didn't notice it being part of the `postgrESql` :O

File: docker_services_cli/config.py
Comment (ppanero) [reply docker_services_cli/config.py:85]: I'll leave it up to @lnielsen, for me it was confusing and I was actually looking for the major definition.

File: docker_services_cli/config.py
Comment (ppanero) [reply docker_services_cli/config.py:52]: For me is fine, it was just a "heads up" :) Change is good!

**Comment by ppanero** [comment]
Discussed IRL, All ok for me :)

File: docker_services_cli/config.py
Comment (diegodelemos) [reply docker_services_cli/config.py]: No need to spend time on this, I've changed it!

File: docker_services_cli/config.py
Comment (diegodelemos) [reply docker_services_cli/config.py]: Let's close this thread, we can leave this for the future, I've left `invenio` for now. If we adopt this in REANA at some point we will make a PR and by this time probably all Invenio's `SQLALCHEMY_DATABASE_URI ` in CI code will be removed.

File: docker_services_cli/env.py
Comment (mvidalgarcia) [inline docker_services_cli/env.py:151]: Moderate: When I run `docker-services-cli down --env` in the output I get

```console
# Configure your environment running:
# eval "$(docker-services-cli down elasticsearch postgresql mysql redis memcached rabbitmq --env)"
```
However,  this `eval` command doesn't work as `down` doesn't receive any options, I think we should eliminate the services list when `cmd` is `down`.

**Review by mvidalgarcia — APPROVED** [review]
Minor: I'm not sure if this is expected but when I install it in editable mode it doesn't work.
```console
$ pip install -e .
Obtaining file:///Users/marco/code/inveniosoftware/docker-services-cli
Collecting click<8.0,>=7.0
  Using cached click-7.1.2-py2.py3-none-any.whl (82 kB)
Installing collected packages: click, Docker-Services-CLI
  Running setup.py develop for Docker-Services-CLI
Successfully installed Docker-Services-CLI click-7.1.2
WARNING: You are using pip version 20.2.4; however, version 20.3.1 is available.
You should consider upgrading via the '/Users/marco/.virtualenvs/docker-services-cli/bin/python -m pip install --upgrade pip' command.
$ docker-services-cli up postgresql12      
Usage: docker-services-cli [OPTIONS] COMMAND [ARGS]...
Try 'docker-services-cli --help' for help.

Error: Invalid value for '--filepath' / '-f': Path '/Users/marco/.virtualenvs/docker-services-cli/lib/python3.8/site-packages/docker_services_cli/docker-services.yml' does not exist.
$ pip install .   
Processing /Users/marco/code/inveniosoftware/docker-services-cli
Requirement already satisfied: click<8.0,>=7.0 in /Users/marco/.virtualenvs/docker-services-cli/lib/python3.8/site-packages (from Docker-Services-CLI==0.2.1) (7.1.2)
Building wheels for collected packages: Docker-Services-CLI
  Building wheel for Docker-Services-CLI (setup.py) ... done
  Created wheel for Docker-Services-CLI: filename=Docker_Services_CLI-0.2.1-py2.py3-none-any.whl size=12266 sha256=a64fa2a771856e6dfe1510edf795649d14ea12525fd4dcc43ce9bfa7f4b5db63
  Stored in directory: /Users/marco/Library/Caches/pip/wheels/2b/3c/85/b3a3a4904a0243297086e24fa3e62ed49e9d00f92f1ba9a5e0
Successfully built Docker-Services-CLI
Installing collected packages: Docker-Services-CLI
  Attempting uninstall: Docker-Services-CLI
    Found existing installation: Docker-Services-CLI 0.2.1
    Uninstalling Docker-Services-CLI-0.2.1:
      Successfully uninstalled Docker-Services-CLI-0.2.1
Successfully installed Docker-Services-CLI-0.2.1
WARNING: You are using pip version 20.2.4; however, version 20.3.1 is available.
You should consider upgrading via the '/Users/marco/.virtualenvs/docker-services-cli/bin/python -m pip install --upgrade pip' command.
$ docker-services-cli up postgresql12
Environment setup
Creating network "docker_services_cli_default" with the default driver
Creating docker_services_cli_postgresql_1 ... done
postgresql up and running!
Services up!
```

I left another command which I think is important but should be easy to fix.
The rest looks fine to me. Approving to avoid blocking the PR... 👍 🚀

**Comment by diegodelemos** [comment]
> Minor: I'm not sure if this is expected but when I install it in editable mode it doesn't work.

I've quickly tested this on `master` branch and it happens there too. Once we merge I will create an issue from your comment to keep track of it.

**Pull request merged**
